contracts/0.4.24/lib/StakeLimitUtils.sol

```diff
@@ -0,0 +1,110 @@
+
+// SPDX-License-Identifier: GPL-3.0
+
+/* See contracts/COMPILERS.md */
+pragma solidity 0.4.24;
+
+//
+// We need to pack four variables into the same 256bit-wide storage slot
+// to lower the costs per each staking request.
+//
+// As a result, slot's memory aligned as follows:
+//
+// LSB ------------------------------------------------------------------------------> MSB
+// 0______________________32______________128_________________________160______________256
+// |______________________|________________|___________________________|________________|
+// | prevStakeBlockNumber | prevStakeLimit | maxStakeLimitGrowthBlocks | maxStakeLimit  |
+// |<----- 32 bits ------>|<-- 96 bits --->|<---------- 32 bits ------>|<--- 96 bits -->|
+//
+//
+// NB: we represent `maxStakeLimitGrowthBlocks` as follows:
+// `maxStakeLimitGrowthBlocks` = `maxStakeLimit` / `stakeLimitIncreasePerBlock`
+//           32 bits                 96 bits               96 bits
+//
+
+library StakeLimitUtils {
+    uint256 internal constant MAX_STAKE_LIMIT_OFFSET = 160;
+    uint256 internal constant MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET = 128;
+    uint256 internal constant PREV_STAKE_LIMIT_OFFSET = 32;
+    uint256 internal constant PREV_STAKE_BLOCK_NUMBER_OFFSET = 0;
+    uint256 internal constant STAKE_LIMIT_PARAMS_MASK = uint256(-1) << MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET;
+
+    /**
+    * @notice Unpack the slot value into stake limit params and state.
+    */
+    function decodeStakeLimitSlot(uint256 _slotValue) internal pure returns (
+        uint96 maxStakeLimit,
+        uint96 stakeLimitIncPerBlock,
+        uint96 prevStakeLimit,
+        uint32 prevStakeBlockNumber
+    ) {
+        maxStakeLimit = uint96(_slotValue >> MAX_STAKE_LIMIT_OFFSET);
+        uint32 growthBlocks = uint32(_slotValue >> MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET);
+        if (growthBlocks > 0) {
+            stakeLimitIncPerBlock = maxStakeLimit / growthBlocks;
+        }
+        prevStakeLimit = uint96(_slotValue >> PREV_STAKE_LIMIT_OFFSET);
+        prevStakeBlockNumber = uint32(_slotValue >> PREV_STAKE_BLOCK_NUMBER_OFFSET);
+    }
+
+    /**
+    * @notice Pack stake limit params and state into a slot.
+    */
+    function encodeStakeLimitSlot(
+        uint96 _maxStakeLimit,
+        uint96 _stakeLimitIncPerBlock,
+        uint96 _prevStakeLimit,
+        uint32 _prevStakeBlockNumber
+    ) internal pure returns (uint256 ret) {
+        ret = uint256(_maxStakeLimit) << MAX_STAKE_LIMIT_OFFSET
+            | uint256(_prevStakeLimit) << PREV_STAKE_LIMIT_OFFSET
+            | uint256(_prevStakeBlockNumber) << PREV_STAKE_BLOCK_NUMBER_OFFSET;
+
+        if (_stakeLimitIncPerBlock > 0) {
+            ret |= uint256(uint32(_maxStakeLimit / _stakeLimitIncPerBlock)) << MAX_STAKE_LIMIT_GROWTH_BLOCKS_OFFSET;
+        }
+    }
+
+    /**
+    * @notice Calculate stake limit for the current block.
+    */
+    function getCurrentStakeLimit(uint256 _slotValue) internal view returns(uint256 limit) {
```

Comment:
Don't you want to call it `calculateCurrentStakeLimit`? It looks like the plain getter now, but it's tricky inside.

Reply:
Hmm, might work! Thank you, gonna change it.